### ISSUE TITLE
state process.absent is failing on Ubuntu 14.04 because psutil is not installed

### ISSUE THREAD

**Issue by wonderslug** [saltstack/salt#19659]
Ran into an issue trying to run process.absent state on Ubuntu 14.04 and worked out that there is no dependency on python-psutil package being installed by default.  This is using the standard ppa install.

This appears to be needed for the for the states.process as well as anything using the modules/ps.py

**Comment by rallytime** [comment]
So specific module dependencies are evaluated on the fly -- so if you install the dependency for a module earlier in the state run, salt will automatically reload the modules afterwards and your specific module will be loaded in time for it to run. We work very hard to keep core dependencies for salt to a minimum, so you might just try using a requisite state to make sure psutil is installed on your minion before running the process.absent state:
```
install_psutil:
  pkg.installed:
    - name: python-psutil
    - require_in:
        - process: your_process_state_id
```

**Comment by wonderslug** [comment]
Ok I can see that  but this seems to be something that is not documented or warning/notices happening when a module is missing.    When running into this case you get a "State 'process.absent' found in SLS 'foo' is unavailable"  In this case the process.absent is a documented core state.  Even when called local with debug there is no log or notice that 

I would say at a minimum that items like this if they are missing a dependency a notice should be raised thats it is unavailable because a dependency is missing.  As it is right now you have no idea why and in this case had to hunt it down in code.

**Comment by basepi** [comment]
The problem is that the loader has no way to tell why a module was not loaded. That is the most common reason a `__virtual__` function would return False, but far from the only one.

That said, I could have sworn we added a note to that error that mentioned checking dependencies. What version of salt are you running?

**Comment by wonderslug** [comment]
Running 2014.7.0

But I looked at the current develop code for modules/ps.py but it only has

```python
# Import third party libs
try:
    import psutil

    HAS_PSUTIL = True
    PSUTIL2 = psutil.version_info >= (2, 0)
except ImportError:
    HAS_PSUTIL = False


def __virtual__():
    if not HAS_PSUTIL:
        return False

    # Functions and attributes used in this execution module seem to have been
    # added as of psutil 0.3.0, from an inspection of the source code. Only
    # make this module available if the version of psutil is >= 0.3.0. Note
    # that this may need to be tweaked if we find post-0.3.0 versions which
    # also have problems running the functions in this execution module, but
    # most distributions have already moved to later versions (for example,
    # as of Dec. 2013 EPEL is on 0.6.1, Debian 7 is on 0.5.1, etc.).
    if psutil.version_info >= (0, 3, 0):
        return True
    return False
```

**Comment by cachedout** [comment]
``The problem is that the loader has no way to tell why a module was not loaded.``

Actually, this isn't true any more. Modules can tell the loader why they didn't load by returning a tuple from ``__virtual__``. Parsing code in load, in ``process_virtual``

**Comment by basepi** [comment]
@wonderslug I'm not saying that missing dependencies are not the problem with `ps`. All I'm saying is that the loader doesn't know why the module was not loaded. All it knows is that `ps.__virtual__` returned `False`, and so the module wasn't loaded.

**Comment by basepi** [comment]
Ah, I found the code to which @cachedout was referring. https://github.com/saltstack/salt/blob/63e207ee302ef4f7164a775f644f0fd02ab15b8a/salt/minion.py#L1109-L1111

However, that code won't be present until 2015.2.0, and modules will have to be modified on a case by case basis to take advantage of it.

**Comment by rallytime** [comment]
@wonderslug I've fixed this in #31878 so that the error message is more helpful:
```
# salt rallytime ps.get_pid_list
rallytime:
    'ps' __virtual__ returned False: The ps module cannot be loaded: psutil not installed.
```
